Shape changing mixin

mhvk opened the pull request:
EDIT: this is now a simplied follow-up of #4096, where a new `ShapedLikeNDArray` mixin class is created, which is used in `Time` and in the various coordinate representations. To also use it in frames and coordinates will require that the frame and coordinate attributes are properly broadcast at creating time.

OLD TEXT FOR REFERENCE
This is a first cut at #4096, where following @embray's suggestion the shape-changing methods from `Time` are split off to a new `ShapedLikeNDArray` mixin class and used in `Coordinates`. With it,
```
In [1]: from astropy.coordinates import SkyCoord; import astropy.units as u, numpy as np

In [2]: sc = SkyCoord(ra=np.arange(9)*u.deg, dec=np.arange(10, 19)*u.deg)

In [3]: sc.reshape(3,3).transpose()
Out[3]: 
<SkyCoord (ICRS): (ra, dec) in deg
    [[(0.0, 10.0), (3.0, 13.0), (6.0, 16.0)],
     [(1.0, 11.0), (4.0, 14.0), (7.0, 17.0)],
     [(2.0, 12.0), (5.0, 15.0), (8.0, 18.0)]]>
```

Comments very welcome (as are PRs with test cases; I'd prefer not to have to write them myself; was boring enough for `Time`!). cc: @taldcroft, @eteq.

Note that there are no test cases yet, though it seems all regular tests pass (which means that `__getitem__` is correctly done). A̶l̶s̶o̶,̶ ̶a̶ ̶p̶r̶o̶b̶l̶e̶m̶ ̶i̶s̶ ̶t̶h̶a̶t̶ ̶`̶s̶c̶.̶T̶`̶ ̶d̶o̶e̶s̶ ̶n̶o̶t̶ ̶w̶o̶r̶k̶ (EDIT: this is solved by the last commit)

mhvk commented:
Note that the sphinx build fails, since somehow the references to `Time.take`, etc., cannot be found any more now they come from the mixin class. Anybody know how to solve this?

embray commented on astropy/utils/misc.py:
```diff
@@ -801,3 +802,96 @@ def __init__(cls, cls_name, bases, members):
+    """Mixin class to provide shape-changing methods.
+
+    These assume that the class proper defines
+    _replicate(method, *args, **kwargs)
```
I guess you can make `_replicate` an abstractmethod?

mhvk replied:
Yes, I thought about making it an ABC as well. It has the advantage that I could also define `ndim` and `shape` as abstract properties, since these are needed for some of the methods.

embray commented:
As for the docs build, I think there's a setting somewhere you can tweak for the API docs for whether or not class methods should be inherited.  I need to check exactly where you change that...

mhvk commented:
@embray - I tried changing `ShapedLikeNDArray` to an abstract base class, as this seemed much more logical, but run into problems with `coordinates.representation.BaseRepresentation` already using a metaclass which is based on `type` (while mine is based on `object`).  It seems difficult to have two metaclasses which are not subclasses of the same class.  Any suggestions?

embray commented:
You'll have to show me what you mean.

mhvk commented:
Fair enough -- see the branch at https://github.com/astropy/astropy/compare/master...mhvk:shape-changing-mixin-abc?expand=1

With that,
```
In [1]: from astropy.coordinates import SkyCoord
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-a44fc1d186bf> in <module>()
----> 1 from astropy.coordinates import SkyCoord

/data/mhvk/venv/astropy-dev/lib/python3.4/site-packages/astropy-1.1.dev13343-py3.4-linux-x86_64.egg/astropy/coordinates/__init__.py in <module>()
     11 from .errors import *
     12 from .angles import *
---> 13 from .baseframe import *
     14 from .distances import *
     15 from .earth import *

/data/mhvk/venv/astropy-dev/lib/python3.4/site-packages/astropy-1.1.dev13343-py3.4-linux-x86_64.egg/astropy/coordinates/baseframe.py in <module>()
     26                      OrderedDescriptorContainer)
     27 from .transformations import TransformGraph
---> 28 from .representation import (BaseRepresentation, CartesianRepresentation,
     29                              SphericalRepresentation,
     30                              UnitSphericalRepresentation,

/data/mhvk/venv/astropy-dev/lib/python3.4/site-packages/astropy-1.1.dev13343-py3.4-linux-x86_64.egg/astropy/coordinates/representation.py in <module>()
     58 
     59 @six.add_metaclass(MetaBaseRepresentation)
---> 60 class BaseRepresentation(ShapedLikeNDArray):
     61     """
     62     Base Representation object, for representing a point in a 3D coordinate

/data/mhvk/venv/astropy-dev/lib/python3.4/site-packages/astropy-1.1.dev13343-py3.4-linux-x86_64.egg/astropy/extern/bundled/six.py in wrapper(cls)
    726             for slots_var in slots:
    727                 orig_vars.pop(slots_var)
--> 728         return metaclass(cls.__name__, cls.__bases__, orig_vars)
    729     return wrapper
    730 

TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```

embray commented:
Oh I see what you're saying.  For that to work you have to make `MetaBaseRepresentation` a subclass of `abc.ABCMeta`. But yeah, it may not be worth the trouble right now.

mhvk commented:
@embray - After a few more trials, it seems it is indeed as simple as making `MetaBaseRepresentation` be a subclass of `abc.ABCMeta` -- and the same for `OrderedDescriptorContainer`. I've now put it here; have a look if you think it makes sense to do it this way.

embray commented:
Yeah, the problem with that now is that everything that uses `OrderedDescriptorContainer` now also becomes an ABC.  Which I think is probably mostly harmless, but it is odd.  This is actually an acknowledged problem with Python metaclasses, and there are a few PEPs in the work to add capabilities to Python classes so that some of the most common problems solved by metaclasses can be accomplished also without them.

But I'm fine with seeing how that goes for now.

mhvk commented:
@embray - I moved the `abc.ABCMeta` inheritance from `OrderedDescriptorContainer` to where it is actually needed. I also tried to solve the issues with sphinx no longer finding the links to `Time.take`, etc., but couldn't. It is hard to experiment, since locally it actually works fine. If you have any suggestion, let me know...

embray commented:
Okay, sounds good.  Odd about the docs.  I'll see if I can reproduce that.

mhvk commented:
@eteq, @taldcroft, @astrofrog - I think enabling multidimensional coordinates and allowing their shape to change would be nice to get in for 1.1, but I don't think I'll get around to adding test cases. Any of you have a change? Obviously, one can use those in `Time` as a basis, to test that views are taken when they are supposed to, etc.

@embray - any luck with figuring out the reason the docs don't link correctly?

taldcroft commented:
@mhvk - sorry, I'm way behind just trying to get all my own PRs and assigned PRs wrapped up.

eteq commented:
@mhvk - I see your point and can try to give it a shoot (especially given its been done for some time...).  

Just so we're clear, it's basically ``time/tests/test_methods.py`` that you're thinking about here, right?

mhvk commented:
@eteq - yes, it is. Not sure one would need to be as thorough as that one, given that the same code is used, but I think it is especially important to check that view and copies are done appropriately (e.g., for the representations, I needed to add `copy=False` in `_replicate`, but I may have missed similar instructions elsewhere, not knowing coordinates all that well).

eteq commented:
sadly, I haven't had time to get tests in on this, and it appears to now need a rebase.  So I think we're going to have to triage it to 1.2 :cry:

mhvk commented:
Yikes, that merge conflict made me realise a change introduced by #3915 that I think is rather in the wrong place (#4222...). Anyway, for now just kept it. Hopefully things still work.

alexrudy commented:
A minor comment: With numpy arrays, it is possible to set the shape attribute to change the shape if that wouldn't change the memory layout of the underlying array (see http://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.shape.html). Perhaps that isn't a guarantee we want to keep, but I use ``somearray.shape = (10,10)`` all the time, and it would be nice if astropy array-like objects allowed that as well.

mhvk commented:
@alexrudy - I did not yet implement shape-setting for coordinates, as it was not something that was as easy to do with the mixin class proposed here as the ndarray methods, which all rely on a `_replicate` method that always returns a new instance. The latter ensures also that coordinate instances remain strictly immutable. It would be quite easy, though, to add shape setting, and personally I think it is a good idea (indeed, for `Time` I had already done so). @eteq - what do you think? It does keep objects immutable in the sense that perhaps matters most, that their data is not changed at all.

bsipocz commented:
The travis docs build failure is unrelated to this PR, the build should pass once restarted.

mhvk commented:
@bsipocz - I checked but the travis build still fails, for the reasons also discussed [above](https://github.com/astropy/astropy/pull/4123#issuecomment-137752262): somehow references to methods of the new shape-changing mixin class are not being followed. Any suggestions?

bsipocz commented:
@mhvk - my bad, should have checked the failing output more carefully. Sadly I have no idea how to fix this and force sphinx to find those methods.

eteq commented:
@mhvk - this now needs rebasing... but if you can do that I will do my best to try to work some docs magic and add tests for this tomorrow.

mhvk commented:
@eteq - I rebased, but stopped the travis tests, since they will fail with my initial tests. The problem is that as currently written the coordinate reshapes all create new arrays -- this will need #4883.

mhvk commented:
@eteq - I'm not sure that without #4883 it is even possible reshape `SkyCoord` objects without making copies.  My preference is to change the way `SkyCoord` behaves, but it may well cause some breakage.  I'm not quite sure how to proceed.

eteq commented:
@mhvk - ok, will have a look at #4883 and see where that goes

mhvk commented:
@eteq - I changed #4883 to make it uncontroversial -- so, I think that can just go in.

I also rebase this on it, since I need to be able to create new coordinate instances without copying the data. With that, the first small tests should now pass, and I'm actually fairly confident the reshaping, etc., will all work. I'm a bit *less* confident all the transformations will continue to work, since I think some of them assume coordinates are arrays of at most 1 dimension. So, that still needs additional tests.

I do hope we can still get this in; I think it is a very nice addition.

eteq commented:
Agreed that we should try to get this in - I'll try to get to a more thorough review of this soonish... or should I give it a bit more if you're planning to make some additional changes, @mhvk?

mhvk commented:
Ah, the tests reminded me that there are still problems with getting the links to the metaclass to show up in sphinx...  Which may not be so easy to solve...

But code-wise, it is all complete. I will on purpose not yet rebase on #4883 to make review easier.

mhvk commented:
I rebased so this doesn't show changes from #4883 any more and cancelled the builds since we know that all but the sphinx one will pass.

eteq commented on astropy/coordinates/tests/test_methods.py:
```diff
@@ -0,0 +1,49 @@
+from ...tests.helper import pytest
+
+
+class TestManipulation():
```
Might it make sense to add a test somewhere (maybe in this file, maybe elsewhere) for higher-dimensional coordinates of some sort?  Like a transformation of a 3x4x2 ICRS coordinate to FK5 with some 3x4x2 array of equinoxes?

eteq commented on astropy/utils/misc.py:
```diff
@@ -850,3 +850,121 @@ def set_locale(name):
+    @abc.abstractproperty
+    def shape(self):
+        """The shape of the instance and underlying arrays."""
+
```
perhaps this should have a concerete ``size`` propertry implemented like this:

```
def size(self):
    return int(np.prod(self.shape))
```
so that then size comes for free even if only `shape` is implemented?

For all implementations above you directly use the ``size`` from the data object, so that's just fine.  But this is an added bonus for possible future subclasses.

mhvk replied:
This is a good idea. I already implemented it locally, and will push it up once I get the time to look into the other attributes of sky coordinates (see comment in main stream).

mhvk replied:
I realised that for `ndim` one also has a sensible default: `len(self.shape)`. So I added that as well.

mhvk replied:
Now also added `isscalar`

mhvk commented:
@eteq - your comment above makes me realise that I was only reshaping the `data` in a coordinate, not any possibly associated attributes. Obviously, like for `Time`, both need to be done, where the general rule may be don't reshape scalars, but everything else is broadcasted against each other and then reshaped.

But for `SkyCoord`, there currently is not a check that inputs have consistent shape.  E.g.,
```
SkyCoord(ra=np.arange(10)*u.deg, dec=np.arange(10)*u.deg, frame='gcrs', obstime=Time([['J2000', 'J2001'], ['J2002', 'J2003']]))
# <SkyCoord (GCRS: obstime=[['J2000.000' 'J2001.000']
#  ['J2002.000' 'J2003.000']], obsgeoloc=[ 0.  0.  0.] m, obsgeovel=[ 0.  0.  0.] m / s): (ra, dec) in deg
#     [(0.0, 0.0), (1.0, 1.0), (2.0, 2.0), (3.0, 3.0), (4.0, 4.0),
#      (5.0, 5.0), (6.0, 6.0), (7.0, 7.0), (8.0, 8.0), (9.0, 9.0)]>
```

Overall, I think this can (and should) be done right, but it will take a bit more time, so I fear this has to be postponed to 1.3.

eteq commented:
@mhvk sad times :( but yes, it's definitely critical that the frame attributes reshape in a way that makes sense, or else a lot of the vectorization breaks.

A lot the tricks using attributes as vectors are somewhat ad-hoc, which is why they aren't well-tested.  But that certainly should change.

mhvk commented:
Indeed, I think it may be best to follow a 3-step process:
1. Separate out the shape-changing routines from `Time` into a mixin-class (done, except that somehow we should make sure the sphinx links work consistently; will start looking at that again once #4924 is merged; right now, building locally doesn't work.)
2. Ensure all components and attributes of `SkyCoord` have consistent dimensions (in `Time`, I ended up deciding this meant a given part either was scalar or was broadcast; just being broadcastable become too much of a hassle to keep track of).
3. Use the shape-mixin in coordinates.

mhvk commented:
Update: I've made the scope of this PR smaller, to only include shape manipulation of representations (where we do not have the issues with other attributes whose shape has to be kept consistent).

mhvk commented:
@eteq - the simplified reshaping, applied just to `Time` and the representations, works well. I'm not quite sure how, but even the sphinx build now passes. I think this can be merged and then further work for coordinates can build on this.

mhvk commented:
@eteq - I think this simplified reshaping PR, which applies just to representations, is all OK now. So, if you have a chance to review...

mhvk commented:
With 1.2 released, it would be nice to move the shape-changing work forward. @eteq, would you have time to review? @StuartLittlefair - this has some interaction with the matrix removal and the general attempts to make `SkyCoord` properly multi-dimensional. Maybe you could have a look?

As I wrote above,  this simplified reshaping PR applies just to representations; `BaseFrame` and `SkyCoord` would build on this.

mhvk commented:
I think this shape-changing PR is now pretty uncontroversial, since it only touches the representations (with a decent suite of tests). I now added a changelog entry. Any chance for the participants in the discussion so far to give it a quick review?

StuartLittlefair commented on astropy/coordinates/representation.py:
```diff
@@ -114,9 +119,22 @@ def get_name(cls):
+        Parameters
+        ----------
+        method : str
+            The method is applied to the internal ``components``.
```
Is it as clear as it could be here that ```method``` must be an attribute of the ```component``` for this to work?

mhvk replied:
Thinking about this more, I think the class docstring could be clearer about what the method should do.

StuartLittlefair commented on astropy/coordinates/representation.py:
```diff
@@ -114,9 +119,22 @@ def get_name(cls):
-    def __getitem__(self, view):
-        return self.__class__(*[getattr(self, component)[view]
-                                for component in self.components])
+    def _replicate(self, method, *args, **kwargs):
```
Thinking about how this might be applied in future, is "replicate" the best name for this function? To me, it implies that the returned ```representation``` will be the same as ```self```, but in fact it is often transformed: for example if the method is ```reshape``` or ```sum``` then the returned ```representation``` is a different shape to the original.

mhvk replied:
@StuartLittlefair - this is a leftover from `Time`, where I hijacked a `replicate` method which in fact did just replicate the instance (without copying the data). I'm quite happy to replace it with something else! Do you have a suggestion?

StuartLittlefair replied:
Perhaps ```apply```?

StuartLittlefair commented on astropy/coordinates/tests/test_representation_methods.py:
```diff
@@ -0,0 +1,122 @@
+        s1_ravel = self.s1.ravel()
+        assert s1_ravel.shape == (self.s1.size,)
+        assert np.all(s1_ravel.lon == self.s1.lon.ravel())
+        assert not np.may_share_memory(s1_ravel.lat, self.s1.lat)
```
for my own education: why are the tests on memory sharing necessary?

mhvk replied:
@StuartLittlefair - in a way, yes, since that was how I found out, e.g., that initially the initialisation of new representation instances made needless copies. For large arrays, I do think it is important that one can guarantee that, e.g., a `reshape` does not make a copy unless it is needed.

StuartLittlefair commented on astropy/utils/misc.py:
```diff
@@ -850,3 +850,134 @@ def set_locale(name):
+        """The shape of the instance and underlying arrays."""
+
+    @abc.abstractmethod
+    def _replicate(method, *args, **kwargs):
```
see comment above about the name of this function.

StuartLittlefair commented on astropy/utils/misc.py:
```diff
@@ -850,3 +850,134 @@ def set_locale(name):
+        return self._replicate('squeeze', *args, **kwargs)
+
+    def take(self, indices, axis=None, mode='raise'):
+        """Return a new instance formed from the elements the given indices.
```
'from the elements at the given indices'

StuartLittlefair commented on astropy/utils/misc.py:
```diff
@@ -850,3 +850,134 @@ def set_locale(name):
+
+    def reshape(self, *args, **kwargs):
+        """Returns an instance containing the same data with a new shape.
+
```
perhaps add 

```python
Refer to :meth:`~numpy.ndarray.reshape` for full documentation
```
as in docstring of ```reshape``` for a numpy array?

mhvk replied:
OK, added a reference to `np.reshape`, since that has the real explanation.

StuartLittlefair commented on astropy/utils/misc.py:
```diff
@@ -850,3 +850,134 @@ def set_locale(name):
+
+    def transpose(self, *args, **kwargs):
+        """Return an instance with the data transposed.
+
```
behaviour is pretty complex for n-D arrays. Should we refer to the numpy documentation for more details?

mhvk replied:
I already refer to the numpy method in the text below, so am a bit confused about what else to do...

StuartLittlefair replied:
Fair enough: I was thinking of copying over some of the relevant material but as you suggest there's probably no need

StuartLittlefair commented on astropy/utils/misc.py:
```diff
@@ -850,3 +850,134 @@ def set_locale(name):
+        else:
+            return self.transpose()
+
+    def swapaxes(self, *args, **kwargs):
```
NumPy's ```swapaxes``` has the signature

```python
a.swapaxes(axis1, axis2)
```

This would probably be clearer to the end user if it had the same signature and docstring. Since ```swapaxes``` only takes two arguments, I don't think theres a use for ```*args``` and ```**kwargs``` here.

mhvk replied:
I thought about copying the signatures, but worried about additions being made; I think overall I prefer to just keep this as is, but can add to the docstring.

StuartLittlefair commented:
Hi @mhvk. I think this looks very nice, though it took me a while to wrap my head around the metaclass magic...

A couple of extra comments:

1. You still define ```ndim```, ```size``` and ```isscalar``` explicitly for ```Time``` objects. That's not necessary any more, is it?

2. Reshaping a ```Time``` object to ```ndim```  greater than 1 breaks ```light_travel_time```, with the following error:

```python
...
/usr/local/lib/python3.5/site-packages/astropy/coordinates/builtin_frames/icrs_cirs_transforms.py in gcrs_to_icrs(gcrs_coo, icrs_frame)
    171     astrom = erfa.apcs13(jd1, jd2, pv)
    172 
--> 173     i_ra, i_dec = aticq(gcrs_ra, gcrs_dec, astrom)
    174 
    175     if gcrs_coo.data.get_name() == 'unitspherical' or gcrs_coo.data.to_cartesian().x.unit == u.one:

/usr/local/lib/python3.5/site-packages/astropy/coordinates/builtin_frames/utils.py in aticq(ri, di, astrom)
    184 
    185     # Bias-precession-nutation, giving GCRS proper direction.
--> 186     ppr = erfa.trxp(astrom['bpn'], pos)
    187 
    188     # Aberration, giving GCRS natural direction

/usr/local/lib/python3.5/site-packages/astropy/_erfa/core.py in trxp(r, p)
  24006 
  24007     #Create the output array, based on the broadcasted shape, adding the generated dimensions if needed
> 24008     broadcast = numpy.broadcast(numpy.int32(0.0), numpy.int32(0.0), r_in[...,0,0], p_in[...,0])
  24009     trp_out = numpy.empty(broadcast.shape + (3,), dtype=numpy.double)
  24010 

ValueError: shape mismatch: objects cannot be broadcast to a single shape
```

It might be best to merge this PR, and create an issue for that problem to be fixed in a separate PR.

mhvk commented:
@StuartLittlefair - indeed, it is not necessary to redefine `ndim`, etc., any more. I had kept them for speed, but on second thought, I think it makes more sense to just use the defaults.

On the `ndim > 1` issue for `Time`: I would think this is already broken in current master too, correct? If so, could you raise a new issue? If not, then I think I should fix it here.

StuartLittlefair commented:
@mhvk - yes it's currently broken in master too. I'll raise a new issue now...

mhvk commented:
@StuartLittlefair - OK, I removed the unnecessary redefinitions of `ndim`, `size` and `isscalar`, and also added a shape setter (plus made that for `Time` safer).

I liked your simple `_apply` alternative to `_replicate`, so use that throughout now.

It should be ready to go in now!

mhvk commented:
@eteq, @taldcroft, @StuartLittlefair - any last comments on the shape-changing PR?

StuartLittlefair commented:
Nothing more to add from me @mhvk!

mhvk commented:
I'm merging this, since it is holding up further work on shape-awareness and broadcasting in coordinates (and therewith fixing, e.g., the bug with light travel times for multidimensional time instances -- #5191).

p.s. the merging travis build will likely fail due to #5198.
